app/src/main/java/com/termux/app/TermuxService.java

```diff
@@ -354,6 +359,23 @@ private void actionReleaseWakeLock(boolean updateNotification) {
         Logger.logDebug(LOG_TAG, "WakeLocks released successfully");
     }
 
+    private void actionServiceStop(Intent intent) {
+        if (intent == null) {
+            Logger.logError(LOG_TAG, "Ignoring null intent to actionServiceStop");
+            return;
+        }
+
+        int gracePeriod = IntentUtils.getIntegerExtraIfSet(intent, TERMUX_SERVICE.EXTRA_TERMINATE_GRACE_PERIOD, 5000);
+
+        Uri executableUri = intent.getData();
+        String executable = UriUtils.getUriFilePathWithFragment(executableUri);
+        String shellName = ShellUtils.getExecutableBasename(executable);
```

Comment:
There would be conflicts if same executable was run from someplace else, and that function only returns first one. You should generate at least a six digit alphanumeric `[A-Z0-9]` random string and store that in cron tab in addition to the id and when sending execute or stop intent, pass that as `TERMUX_SERVICE.EXTRA_SHELL_NAME`, so that the executable for the job is killed, and not any other.

https://github.com/termux/termux-app/commit/d287734abae95107e73c763ccbd807315ca9a9c0

https://github.com/termux/termux-app/issues/3709

Reply:
Good point.
Is there a character limit on the shell name?
Or the other way around: Could I just use a UUID?

Reply:
There is no limit currently, but there should be, ideally `255`, equal to `NAME_NAME`.

UUID would be too long and harder to see in app shells UI list and logs, just use a six character alphanumeric `[a-zA-Z0-9]` string, like `mktemp` template uses for unique files. Following should work, make sure `<script_name>_<public_id>_<private_id>` doesn't already exist in cron tab, otherwise regenerate `private_id`. The `public_id` here is the incrementing number you are currently using.

```java
char[] allowedCharsArray = ("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789").toCharArray();
char[] private_id = new char[6];
Random random = new SecureRandom();
for (int i = 0; i < private_id.length; i++) {
    private_id[i] = allowedCharsArray[random.nextInt(allowedCharsArray.length)];
}
```

https://man7.org/linux/man-pages/man3/mktemp.3.html

https://bugs.python.org/issue12015

Reply:
I see. Thanks for the example.

I changed the method to only work if the shell name is explicitly set.

Reply:
Welcome. Makes sense. I think it would be better to call `getTermuxTaskForShellName()` in a `while` loop until it starts returning `null`, so that all shells with same name get killed, since currently only first one would get killed, which may not be the one caller intended to kill. To prevent duplicates, callers should start shells with a unique shell name, like the cron API would do.

Reply:
Good point - I had not considered multiple shells with the same name.